### ISSUE TITLE
Utf-8 for minetest 0.5

### ISSUE THREAD

**Issue by HybridDog** [minetest/minetest#6938]
I think, since the pending minetest 0.5 is incompatible to 0.4.x, you could switch to Utf-8 without compatibility code.
There're already an "Adoption needed" PR for utf-8 chat: https://github.com/minetest/minetest/pull/2885
Now it's the time for compat breaking PRs, isn't it?

**Comment by NewbProgrammer101** [comment]
Indeed it is. :)

**Comment by nerzhul** [comment]
@HybridDog it can be the moment, but what are the side effects on mods ?

**Comment by Fixer-007** [comment]
Will also help for cyrillics in text I guess (it does not work fully atm)

**Comment by benrob0329** [comment]
@nerzhul Lua5.3 has extra functions for UTF8 usage.

**Comment by red-001** [comment]
we can't update to it easily, but I suppose we could try and backport them.

**Comment by Wuzzy2** [comment]
I don't understand why mods should break because of this. o_O

**Comment by HybridDog** [comment]
I haven't seen a mod yet which relies on utf-16.

**Comment by 4w** [comment]
> but what are the side effects on mods ?

*[UTF-8] was designed for backward compatibility with ASCII. Code points with lower numerical values, which tend to occur more frequently, are encoded using fewer bytes. The first 128 characters of Unicode, which correspond one-to-one with ASCII, are encoded using a single octet with the same binary value as ASCII, so that valid ASCII text is valid UTF-8-encoded Unicode as well.* [Wikipedia](https://en.wikipedia.org/wiki/UTF-8)

If I understand that correctly none of the texts has to be changed and all mods should work. It should just made sure that all text-related API functions properly use UTF-8 which makes them not only UTF-8 compatible but will still work with ASCII.

**Comment by HybridDog** [comment]
There are actually only very few mods which use anything other than ascii, e.g. translation mods (intllib).
And they use Utf-8 I think.

**Comment by raymoo** [comment]
Don't know if it's an irrlicht issue, but support for IME input in X11 or at least clipboard pasting support would be nice. Currently if you paste non-ascii unicode characters from X11 they come out as a bunch of escape codes like "\uXXXX\uYYYY\uZZZZ" etc.

**Comment by Lourens-Rich** [comment]
Also we needed in full UTF-8 support for signs. (I'm about copy and paste features)

**Comment by MuhdNurHidayat** [comment]
Currently if you paste non-ascii unicode characters in Minetest chat running on Ubuntu, it appears as escape codes too, and on Windows it appears as random ascii characters...

So support for proper UTF-8 copy/paste, or including support of IME input would be nice.

**Comment by MuhdNurHidayat** [comment]
Reading Irrlicht's forum linked above, they wrote it should be possible to copy paste, so here the problem should be in the implementation.

----
Oh yeah, I remembered that I could chat in languages that use IME just fine on official Android port but the formspecs can't support them. Basically, I could chat using IME on Android Minetest where the text is passed directly to the game, but I can't use IME on formspecs (eg. search box at inventory).

On the other hand, Ubuntu (on my installation) and Windows can't accept IME at all even in chat.

Since IME is just a layer of software which processes text input before passing the real UTF-8 output to the program, it should not be impossible to implement something that allow IME displayed layered or hovering ingame and let the game wait the IME process the text before accept the UTF-8 output from IME... but that's a separate issue I guess (#3022).

**Comment by numberZero** [comment]
From a bug report:

> When you insert a text into the Lua Microcontroller, select it and copy it into your clipboard with Ctrl+C, all multibyte characters are broken, as only the last byte is copied.
>
> For example, if you insert the character "ł" (on my keyboard this is AltGr + L), it becomes "B" (U+0042 instead of U+0142).

Smells like downcasting `wchar_t` to `char` somewhere...

**Comment by Wuzzy2** [comment]
Just for the record:
As a modder myself, I fully support the proposed (proper!) switch to UTF-8.
I am willing to face potential breakage in my mods, although I doubt it will be a huge problem.

Going UTF-8 is worth the risk. Continuing with the half-broken text support we have now is a much worse option.

**Comment by SmallJoker** [comment]
@HybridDog Lua does not use any encoding. It's just treating strings as byte arrays.

**Comment by HybridDog** [comment]
The lua source files, where modders can put strings with unicode characters, are encoded in UTF-8 and not UTF-16 as far as I know.

**Comment by numberZero** [comment]
@HybridDog Lua is encoding-agnostic. You may use encodings like UTF-8, Latin-1, Windows-949 or even EBCDIC (carefully) for string literals and comments; only executable code itself must be valid ASCII. UTF-16 (and its subset UCS-2) is not allowed as it conflicts hardly with ASCII; e.g. it has NUL byte in representation of any basic character. It may be stored in Lua strings by other means, though, as they are NUL-safe (unlike C strings).

**Comment by paramat** [comment]
1 year bump, any core dev support?

Does this actually need to be decided before 5.0.0?

**Comment by srifqi** [comment]
I second with @HybridDog and @Wuzzy2. It is worth the risk and this is the perfect time to do this.

**Comment by numberZero** [comment]
@paramat “Can’t add?” Why?

**Comment by rubenwardy** [comment]
I guess because 5.0 has been release :man_shrugging:

**Comment by Wuzzy2** [comment]
So UTF-8 support will never be added?

**Comment by paramat** [comment]
Yes because too late for 5.0.0. But also no coredev support.
Could change the issue title and reopen for adding while dealing with incompatibility.

**Comment by Wuzzy2** [comment]
Maybe UTF-8 support will be added in Minetest 6.0.0? XD

**Comment by HybridDog** [comment]
You could add a Compatibility Break label to this issue.

**Comment by rubenwardy** [comment]
I don't see why this needs a compatibility break. We survived for 7 years by using this amazing thing called a protocol version

**Comment by HybridDog** [comment]
Sorry, I mean the Protocol bump label.

**Comment by sfan5** [comment]
#10879 and #10881 should fix many of the remaining issues with UTF-8 vs. WCHAR vs. UTF-16 or 32 on different platforms.

If there are separate issues with e.g. copy-paste I suggest opening extra issues for them (if not open already).
